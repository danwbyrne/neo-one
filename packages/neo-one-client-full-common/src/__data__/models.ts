import { common, ContractParameterTypeModel, ECPoint, SignatureString, UInt160 } from '@neo-one/client-common';
import { constants } from '@neo-one/utils';
import {
  ContractABIModel,
  ContractEventModel,
  ContractFunctionModel,
  ContractGroupModel,
  ContractManifestModel,
  ContractModel,
  ContractParameterModel,
  ContractPermissionDescriptorModel,
  ContractPermissionsModel,
  ContractPropertyStateModel,
} from '../models';

export const contractParamModel = {
  boolean: new ContractParameterModel({
    type: ContractParameterTypeModel.Boolean,
    name: 'param',
  }),
  byteArray: new ContractParameterModel({
    type: ContractParameterTypeModel.ByteArray,
    name: 'param',
  }),
  hash160: new ContractParameterModel({
    type: ContractParameterTypeModel.Hash160,
    name: 'param',
  }),
  hash256: new ContractParameterModel({
    type: ContractParameterTypeModel.Hash256,
    name: 'param',
  }),
  array: new ContractParameterModel({
    type: ContractParameterTypeModel.Array,
    name: 'param',
  }),
  integer: new ContractParameterModel({
    type: ContractParameterTypeModel.Integer,
    name: 'param',
  }),
  interopInterface: new ContractParameterModel({
    type: ContractParameterTypeModel.InteropInterface,
    name: 'param',
  }),
  map: new ContractParameterModel({
    type: ContractParameterTypeModel.Map,
    name: 'param',
  }),
  publicKey: new ContractParameterModel({
    type: ContractParameterTypeModel.PublicKey,
    name: 'param',
  }),
  signature: new ContractParameterModel({
    type: ContractParameterTypeModel.Signature,
    name: 'param',
  }),
  string: new ContractParameterModel({
    type: ContractParameterTypeModel.String,
    name: 'param',
  }),
  void: new ContractParameterModel({
    type: ContractParameterTypeModel.Void,
    name: 'param',
  }),
};

export const contractFunctionModel = (
  parameters: readonly ContractParameterModel[] = [contractParamModel.boolean],
  returnType: ContractParameterTypeModel = ContractParameterTypeModel.Void,
  name = 'function',
) => new ContractFunctionModel({ name, parameters, returnType });

export const contractEventModel = (
  parameters: readonly ContractParameterModel[] = [contractParamModel.boolean],
  name = 'event',
) => new ContractEventModel({ name, parameters });

export const contractAbiModel = (
  methods: readonly ContractFunctionModel[] = [contractFunctionModel()],
  events: readonly ContractEventModel[] = [contractEventModel()],
  entryPoint: ContractFunctionModel = contractFunctionModel(),
  hash = common.bufferToUInt160(Buffer.alloc(20, 1)),
) => new ContractABIModel({ hash, entryPoint, methods, events });

export const contractGroupModel = (
  publicKey: ECPoint = common.stringToECPoint(constants.PRIVATE_NET_PUBLIC_KEY),
  signature: SignatureString = 'ccaab040cc25021c91567b75db4778853441869157b8f6aad960cdcf1069812480027a528ca9b98e2205027de20696f848cf81824eeb7af1d5110870870ceb67',
) => new ContractGroupModel({ publicKey, signature });

export const contractPermissionDescriptorModel = (hashOrGroupType: 'uint160' | 'ecpoint' | undefined) => {
  let hashOrGroup: UInt160 | ECPoint | undefined;
  if (hashOrGroupType === 'uint160') {
    hashOrGroup = common.bufferToUInt160(Buffer.alloc(20, 1));
  } else if (hashOrGroupType === 'ecpoint') {
    hashOrGroup = common.stringToECPoint(constants.PRIVATE_NET_PUBLIC_KEY);
  }

  return new ContractPermissionDescriptorModel({ hashOrGroup });
};

export const contractPermissionsModel = (
  hashOrGroupType: 'uint160' | 'ecpoint' | undefined,
  methods: readonly string[] = [],
) => new ContractPermissionsModel({ contract: contractPermissionDescriptorModel(hashOrGroupType), methods });

export const contractManifestModel = (
  groups: readonly ContractGroupModel[] = [contractGroupModel()],
  features: ContractPropertyStateModel = ContractPropertyStateModel.HasStoragePayable,
  abi: ContractABIModel = contractAbiModel(),
  permissions: readonly ContractPermissionsModel[] = [contractPermissionsModel('uint160', ['method1'])],
  trusts: readonly UInt160[] = [common.bufferToUInt160(Buffer.alloc(20, 1))],
  safeMethods: readonly string[] = ['method1', 'method2'],
) => new ContractManifestModel({ groups, features, abi, permissions, trusts, safeMethods });

export const contractModel = (
  script: Buffer = Buffer.alloc(25),
  manifest: ContractManifestModel = contractManifestModel(),
) => new ContractModel({ script, manifest });
