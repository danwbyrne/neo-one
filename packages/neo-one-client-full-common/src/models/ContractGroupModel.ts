import {
  BinaryWriter,
  createSerializeWire,
  ECPoint,
  SerializableWire,
  SerializeWire,
  SignatureString,
} from '@neo-one/client-common';
import { BaseState } from './BaseState';

export interface ContractGroupModelAdd {
  readonly publicKey: ECPoint;
  readonly signature: SignatureString;
}

export class ContractGroupModel extends BaseState implements SerializableWire<ContractGroupModel> {
  public readonly publicKey: ECPoint;
  public readonly signature: SignatureString;
  public readonly serializeWire: SerializeWire = createSerializeWire(this.serializeWireBase.bind(this));

  public constructor({ publicKey, signature }: ContractGroupModelAdd) {
    super({ version: undefined });
    this.publicKey = publicKey;
    this.signature = signature;
  }

  public serializeWireBase(writer: BinaryWriter): void {
    serializeContractGroupWireBase({ writer, group: this });
  }
}

export const serializeContractGroupWireBase = ({
  writer,
  group,
}: {
  readonly writer: BinaryWriter;
  readonly group: ContractGroupModel;
}): void => {
  writer.writeECPoint(group.publicKey);
  writer.writeVarString(group.signature);
};
