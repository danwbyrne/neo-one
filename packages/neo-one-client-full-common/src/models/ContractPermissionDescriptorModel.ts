import {
  BinaryWriter,
  common,
  createSerializeWire,
  ECPoint,
  SerializableWire,
  SerializeWire,
  UInt160,
} from '@neo-one/client-common';
import { BaseState } from './BaseState';

export interface ContractPermissionDescriptorModelAdd {
  readonly hashOrGroup: UInt160 | ECPoint | undefined;
}

export class ContractPermissionDescriptorModel extends BaseState
  implements SerializableWire<ContractPermissionDescriptorModel> {
  public readonly hashOrGroup: UInt160 | ECPoint | undefined;
  public readonly serializeWire: SerializeWire = createSerializeWire(this.serializeWireBase.bind(this));

  public constructor({ hashOrGroup }: ContractPermissionDescriptorModelAdd) {
    super({ version: undefined });
    this.hashOrGroup = hashOrGroup;
  }

  public serializeWireBase(writer: BinaryWriter): void {
    serializeContractPermissionDescriptorWireBase({ writer, permission: this });
  }
}

export const serializeContractPermissionDescriptorWireBase = ({
  writer,
  permission,
}: {
  readonly writer: BinaryWriter;
  readonly permission: ContractPermissionDescriptorModel;
}): void => {
  const { hashOrGroup } = permission;

  if (hashOrGroup === undefined) {
    writer.writeVarBytesLE(Buffer.from([]));
  } else if (common.isUInt160(hashOrGroup)) {
    writer.writeUInt160(hashOrGroup);
  } else if (common.isECPoint(hashOrGroup)) {
    writer.writeECPoint(hashOrGroup);
  }
};
