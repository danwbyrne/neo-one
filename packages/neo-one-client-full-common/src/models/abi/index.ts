export * from './ContractABIModel';
export * from './ContractParameterModel';
export * from './ContractFunctionModel';
export * from './ContractEventModel';
