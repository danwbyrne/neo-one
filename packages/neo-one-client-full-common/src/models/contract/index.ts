export * from './ContractManifestModel';
export * from './ContractStateModel';
export * from './ContractModel';
export * from './abi';
export * from './permissions';
