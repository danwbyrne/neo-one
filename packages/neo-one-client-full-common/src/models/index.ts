export * from './abi';
export * from './BaseState';
export * from './ContractGroupModel';
export * from './ContractManifestModel';
export * from './ContractPermissionDescriptorModel';
export * from './ContractPermissionsModel';
export * from './ContractModel';
export * from './ContractPropertyStateModel';
export * from './transaction';
