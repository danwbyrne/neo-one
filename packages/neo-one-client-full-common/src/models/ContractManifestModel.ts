import {
  BinaryWriter,
  common,
  createSerializeWire,
  SerializableWire,
  SerializeWire,
  UInt160,
  UInt160Hex,
  utils,
} from '@neo-one/client-common';
import { ContractABIModel } from './abi';
import { BaseState } from './BaseState';
import { ContractGroupModel } from './ContractGroupModel';
import { ContractPermissionsModel } from './ContractPermissionsModel';
import { ContractPropertyStateModel, HasPayable, HasStorage } from './ContractPropertyStateModel';

export interface ContractManifestModelAdd {
  readonly groups: readonly ContractGroupModel[];
  readonly features: ContractPropertyStateModel;
  readonly abi: ContractABIModel;
  readonly permissions: readonly ContractPermissionsModel[];
  readonly trusts: readonly UInt160[];
  readonly safeMethods: readonly string[];
}

export class ContractManifestModel extends BaseState implements SerializableWire<ContractManifestModel> {
  public get hash(): UInt160 {
    return this.hashInternal();
  }

  public get hashHex(): UInt160Hex {
    return this.hashHexInternal();
  }
  public readonly maxLength = 2048;
  public readonly abi: ContractABIModel;
  public readonly groups: readonly ContractGroupModel[];
  public readonly permissions: readonly ContractPermissionsModel[];
  public readonly trusts: readonly UInt160[];
  public readonly safeMethods: readonly string[];
  public readonly hasStorage: boolean;
  public readonly payable: boolean;
  public readonly features: ContractPropertyStateModel;
  public readonly serializeWire: SerializeWire = createSerializeWire(this.serializeWireBase.bind(this));
  private readonly hashInternal = utils.lazy(() => this.abi.hash);
  private readonly hashHexInternal = utils.lazy(() => common.uInt160ToHex(this.hash));

  public constructor({ abi, groups, features, permissions, trusts, safeMethods }: ContractManifestModelAdd) {
    super({ version: undefined });
    this.abi = abi;
    this.groups = groups;
    this.permissions = permissions;
    this.trusts = trusts;
    this.safeMethods = safeMethods;
    this.features = features;
    this.hasStorage = HasStorage.has(features);
    this.payable = HasPayable.has(features);
  }

  public serializeWireBase(writer: BinaryWriter): void {
    serializeContractManifestWireBase({ writer, manifest: this });
  }
}

export const serializeContractManifestWireBase = ({
  writer,
  manifest,
}: {
  readonly writer: BinaryWriter;
  readonly manifest: ContractManifestModel;
}): void => {
  manifest.abi.serializeWireBase(writer);
  manifest.groups.forEach((group) => group.serializeWireBase(writer));
  manifest.permissions.forEach((permission) => permission.serializeWireBase(writer));
  manifest.trusts.forEach((trust) => writer.writeUInt160(trust));
  manifest.safeMethods.forEach((method) => writer.writeVarString(method));
  writer.writeUInt8(manifest.features);
};
