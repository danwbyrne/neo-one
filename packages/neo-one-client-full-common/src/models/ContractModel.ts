import { BinaryWriter, createSerializeWire, SerializableWire, SerializeWire } from '@neo-one/client-common';
import { BaseState } from './BaseState';
import { ContractManifestModel } from './ContractManifestModel';

export interface ContractModelAdd {
  readonly script: Buffer;
  readonly manifest: ContractManifestModel;
}

export class ContractModel extends BaseState implements SerializableWire<ContractModel> {
  public readonly script: Buffer;
  public readonly manifest: ContractManifestModel;
  public readonly serializeWire: SerializeWire = createSerializeWire(this.serializeWireBase.bind(this));

  public constructor({ script, manifest }: ContractModelAdd) {
    super({ version: undefined });
    this.script = script;
    this.manifest = manifest;
  }

  public serializeWireBase(writer: BinaryWriter): void {
    serializeContractWireBase({ writer, contract: this });
  }
}

export const serializeContractWireBase = ({
  writer,
  contract,
}: {
  readonly writer: BinaryWriter;
  readonly contract: ContractModel;
}): void => {
  writer.writeVarBytesLE(contract.script);
  contract.manifest.serializeWireBase(writer);
};
