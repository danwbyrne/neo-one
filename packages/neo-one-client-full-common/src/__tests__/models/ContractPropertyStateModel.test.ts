import { assertContractPropertyState, getContractProperties, HasPayable, HasStorage } from '../../models';

describe('ContractPropertyStateModel', () => {
  test('isContractPropertyState', () => {
    expect(assertContractPropertyState(0)).toEqual(0);
    expect(assertContractPropertyState(1)).toEqual(1);
    expect(assertContractPropertyState(4)).toEqual(4);
    expect(assertContractPropertyState(5)).toEqual(5);

    const assertThrowNeg1 = () => assertContractPropertyState(-1);
    expect(assertThrowNeg1).toThrowErrorMatchingSnapshot();
    const assertThrow2 = () => assertContractPropertyState(2);
    expect(assertThrow2).toThrowErrorMatchingSnapshot();
    const assertThrow3 = () => assertContractPropertyState(3);
    expect(assertThrow3).toThrowErrorMatchingSnapshot();
    const assertThrow6 = () => assertContractPropertyState(6);
    expect(assertThrow6).toThrowErrorMatchingSnapshot();
  });

  test('getContractProperties', () => {
    expect(getContractProperties({ hasStorage: false, payable: false })).toEqual(0);
    expect(getContractProperties({ hasStorage: false, payable: true })).toEqual(4);
    expect(getContractProperties({ hasStorage: true, payable: false })).toEqual(1);
    expect(getContractProperties({ hasStorage: true, payable: true })).toEqual(5);
  });

  test('HasStorage', () => {
    expect(HasStorage.has(0)).toEqual(false);
    expect(HasStorage.has(1)).toEqual(true);
    expect(HasStorage.has(2)).toEqual(false);
    expect(HasStorage.has(3)).toEqual(false);
    expect(HasStorage.has(4)).toEqual(false);
    expect(HasStorage.has(5)).toEqual(true);
    expect(HasStorage.has(6)).toEqual(false);
  });

  test('HasPayable', () => {
    expect(HasPayable.has(0)).toEqual(false);
    expect(HasPayable.has(1)).toEqual(false);
    expect(HasPayable.has(2)).toEqual(false);
    expect(HasPayable.has(3)).toEqual(false);
    expect(HasPayable.has(4)).toEqual(true);
    expect(HasPayable.has(5)).toEqual(true);
    expect(HasPayable.has(6)).toEqual(false);
  });
});
