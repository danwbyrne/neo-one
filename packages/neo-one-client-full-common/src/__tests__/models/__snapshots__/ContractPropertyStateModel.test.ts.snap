// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContractPropertyStateModel isContractPropertyState 1`] = `"Expected contract parameter type, found: -1"`;

exports[`ContractPropertyStateModel isContractPropertyState 2`] = `"Expected contract parameter type, found: 2"`;

exports[`ContractPropertyStateModel isContractPropertyState 3`] = `"Expected contract parameter type, found: 3"`;

exports[`ContractPropertyStateModel isContractPropertyState 4`] = `"Expected contract parameter type, found: 6"`;
